Feature/multi tenancy

salmanazmat666 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR is only meant to have a discussion around the implementation of multi-tenancy in oauth2 proxy and have comments from the maintainers about if they'll be willing to merge when the feature is fully implemented and tested. I have tested the basic Oauth2 flow with OIDC type provider and it is working as expected.
<!--- Describe your changes in detail -->

## Motivation and Context
In its current form, oauth2 proxy cannot handle multiple tenants, each with different IdP configurations and requirements. Multi-tenancy is a very common design pattern in SAAS applications and issues have been raised to implement this feature on oauth2 proxy in the past like #1863 and #739. We can be sure that the requirement for this feature will increase in future.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
This PR is not ready to merge in its current form, and tests are broken right now.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Implementation Details

Following are the key points that'll help understand the implementation

- Following interface has been introduced, which is responsible for loading a Tenant given its `id` or `*http.Request`
```
type Loader interface {
	Load(req *http.Request) (*Tenant, error)
	LoadById(id string) (*Tenant, error)
}
type Tenant struct {
	Id       string
	Provider providers.Provider
}
```
- Instead of using the provider configured in the `OAuthProxy` struct, now the `Tenant` struct is loaded in a middleware using the `Loader` interface and injected into the request context. Later in the flow, the `Tenant` can be retrieved from the request context wherever needed.
- One assumption is that the incoming request contains all the information to choose the corresponding tenant. We can have tenantId as part of the host url, query parameters, path or headers.
- Currently, two implementations have been implemented for Tenant `Loader` interface. 
  - One, that loads the tenants from a config file and doesn't allow to add/remove tenants dynamically. This implementation has following sample configuration
   ```
   tenantLoader:
     type: "config"
     config:
       tenantsFile: "tenants.yaml" // this file contains tenants' configuration,
       rules:
         - source: "header"
           header: "Tenantid"
           expr: (^[^\.]*) # matches everything before the first '.', so gets 'www' out from 'www.google.com'
           captureGroup: 1
         - source: "host"
           expr: (^[^\.]*) # matches everything before the first '.', so gets 'www' out from 'www.google.com'
           captureGroup: 1
   ```
   This implementation of tenant `Loader` allows you to describe a list of rules based on which we can extract tenant id from the request. Rule configuration is as follows
   ```
   type RuleSource string

   const (
      SourceHost        RuleSource = "host"
      SourcePath        RuleSource = "path"
      SourceQueryParams RuleSource = "query"
      SourceHeader      RuleSource = "header"
   )

   type RuleConfig struct {
     Source       RuleSource // which part of the HTTP request contains the tenant id
     Expr         string     // the regex expression to match and extract tenant id from the source
     CaptureGroup int        // capture group or sub-match that is actually the tenant id
     QueryParam   string     // query parameter in case source is 'query'
     Header       string     // header key in case source is 'header'
   }
   ```
  - Second that loads a single pre-fixed tenant no matter what id we use in `loader.LoadById(string)` method or what request we pass to `loader.Load(*http.Request)` method. This second implementation allows us to use the legacy config.
- Using the `Loader` interface we can implement the the functionality that loads/removes the tenants dynamically, without actually touching the actual oauth2 proxy logic.
-  Alpha config feature is used to configure the tenant `Loader` interface, and everything is backwards compatible so legacy config can be used as well without configuring the tenant `Loader`. If the legacy config is used, we configure the tenant `Loader` interface and initialize it using the second implementation described above, in the `LegacyOptions.ToOptions()` method. 

## TODOs

1. Implement new tests and update the broken ones.
2. Update Documentation.
3. Implement multi-tenancy in following constructs since they're also tenant specific
    - Email domains and authenticated emails validator
    - Trusted IPs validator
    - Basic Auth validator
    - SkipJwtBearerTokens option
4. Other things, I might have missed

## Final Comments

There is still lots of work to make the oauth2 proxy fully multi-tenant but we can take on things one by one which is possible because we are using alpha config which still allows for breaking changes.

This is really a great work by all the contributors and I want to thank you guys for this. Even if this doesn't get approved we'll be forking and implementing the features privately anyway.

Looking forward to hear back from maintainers. @JoelSpeed @NickMeves

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

Commit:
add tenantLoader interface and it's implementations

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

Commit:
use tenantloader to add multi tenant capability

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

Commit:
add alpha-config and tenantsConfig sample files

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

zeeshanshabbir93 commented:
Hi, @JoelSpeed   Are you guys interested in adding this feature to oauth2proxy anytime soon? We plan on using implementation in this PR for our internal use but we have put in extra effort to make it backward compatible  & generic as we want to contribute to the community. But if this doesn't look like something that you would want to merge then we can take a private fork & maintain a custom implementation without the overhead of backward compatibility etc.
Looking forward to your response.

JoelSpeed commented:
We are interested in this feature, but, I'm not getting much time for the project lately.

I went travelling for a few months this year and am currently mid way through a house move.

I expect I'll have time to come to this sometime mid to late January once things have settled down in my life a little bit. Apologies for the delays folks, but please do remember, this is a passion project for me, I don't get time day to day for it

JoelSpeed commented on pkg/apis/options/alpha_options.go:
```diff
@@ -42,7 +47,9 @@ type AlphaOptions struct {
 
 	// Providers is used to configure multiple providers.
-	Providers Providers `json:"providers,omitempty"`
+	Providers provideropts.Providers `json:"providers,omitempty"`
```
Why has this been moved into a new package?

salmanazmat666 replied:
otherwise get import cycles,

```
package github.com/oauth2-proxy/oauth2-proxy/v7
        imports github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options
        imports github.com/oauth2-proxy/oauth2-proxy/v7/tenant
        imports github.com/oauth2-proxy/oauth2-proxy/v7/tenant/configloader
        imports github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options: import cycle not allowed
```

basically ConfigLoader (a TenantLoader implementation) config depends on the provider config, which is part of `pkg/apis/options` package, and the options.Options depend on the TenantLoader config which depends upon the ConfigLoader config.

This is resolved by moving the provider options to a new package.

salmanazmat666 replied:
> Why has this been moved into a new package?

resolved. moved to the original location.

salmanazmat666 replied:
> Why has this been moved into a new package?

resolved. moved to the original location.

JoelSpeed commented on pkg/apis/options/alpha_options.go:
```diff
@@ -42,7 +47,9 @@ type AlphaOptions struct {
-	Providers Providers `json:"providers,omitempty"`
+	Providers provideropts.Providers `json:"providers,omitempty"`
+
+	TenantLoader tenant.LoaderConfiguration `flag:"tenant-loader" cfg:"tenant_loader"`
```
This struct is meant to represent the configuration, please don't add runtime bits to it

salmanazmat666 replied:
there _are_ configurations, not runtime structures,

salmanazmat666 replied:
resolved. removed the runtime bits from options.

JoelSpeed commented on pkg/apis/options/legacy_options.go:
```diff
@@ -93,6 +97,14 @@ func (l *LegacyOptions) ToOptions() (*Options, error) {
 	}
 	l.Options.Providers = providers
 
+	// in case of legacy options, we initilize single tenant implementation of tenant loader
+	l.Options.TenantLoader = tenant.LoaderConfiguration{
+		Type: "single",
+		Single: &single.Configuration{
+			Provider: &providers[0],
+		},
+	}
```
This kind of initialisation shouldn't be happening within the options part of the project

salmanazmat666 replied:
agreed,

salmanazmat666 replied:
resolved, removed the initialization logic away from the `options` package.

JoelSpeed commented on pkg/apis/options/loginurlopts/login_url_parameters.go:
```diff
@@ -1,4 +1,4 @@
-package options
+package loginurlopts
```
Why has this moved?

salmanazmat666 replied:
> otherwise get import cycles,
> 
> ```
> package github.com/oauth2-proxy/oauth2-proxy/v7
>         imports github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options
>         imports github.com/oauth2-proxy/oauth2-proxy/v7/tenant
>         imports github.com/oauth2-proxy/oauth2-proxy/v7/tenant/configloader
>         imports github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options: import cycle not allowed
> ```
> 
> basically ConfigLoader (a TenantLoader implementation) config depends on the provider config, which is part of `pkg/apis/options` package, and the options.Options depend on the TenantLoader config which depends upon the ConfigLoader config.
> 
> This is resolved by moving the provider options to a new package.

same thing as above, since provider options depends on loginurlopts,
we can keep the provider opts and loginurlopts in the same package though,

salmanazmat666 replied:
resolved, moved the file/s back to its original location.

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -40,6 +42,9 @@ type Options struct {
 	// Not used in the legacy config, name not allowed to match an external key (upstreams)
 	// TODO(JoelSpeed): Rename when legacy config is removed
+
+	TenantLoader tenant.LoaderConfiguration `cfg:",internal"`
```
Please don't add runtime information to structs that are intended for configuration

salmanazmat666 replied:
this is the configuration for tenantLoader, not the runtime construct,

salmanazmat666 replied:
resolved, remved the runtime bits away from the options package.

JoelSpeed commented on pkg/apis/options/provideropts/providers.go:
```diff
@@ -1,4 +1,6 @@
-package options
+package provideropts
```
Why is this moving?

salmanazmat666 replied:
due to import cycles, as described [here](https://github.com/oauth2-proxy/oauth2-proxy/pull/1923#discussion_r1060519964),

salmanazmat666 replied:
resolved. moved to original location.

JoelSpeed commented on pkg/app/pagewriter/pagewriter.go:
```diff
@@ -3,14 +3,16 @@ package pagewriter
 	"fmt"
 	"net/http"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/tenant/types"
```
Do we need another concept, a tenant? Why not extend the provider interface?

salmanazmat666 replied:
responded [here](https://github.com/oauth2-proxy/oauth2-proxy/pull/1923#issuecomment-1369665582),

salmanazmat666 replied:
the tenant struct and loader interfaces have been removed. Only a concep of tenant-id remains now. We extract tenant-id from the incoming request. We could have replaced tenant-id as well and only captured provider-id from the request. But in order to make the whole module multi-tenant, we must have the tenant-concept in some form. Since there are other components besides providers that need to be modified to make them multi-tenant. For example, basicAuth validator, email domain validator and SkipJwtBearerTokens option.

JoelSpeed commented on pkg/middleware/stored_session.go:
```diff
@@ -40,14 +42,6 @@ type StoredSessionLoaderOptions struct {
 	// How often should sessions be refreshed
 	RefreshPeriod time.Duration
-
-	// Provider based session refreshing
-	RefreshSession func(context.Context, *sessionsapi.SessionState) (bool, error)
-
-	// Provider based session validation.
-	// If the sesssion is older than `RefreshPeriod` but the provider doesn't
-	// refresh it, we must re-validate using this validation.
-	ValidateSession func(context.Context, *sessionsapi.SessionState) bool
```
This seems like a good idea 🤔

salmanazmat666 replied:
would you elaborate please.
I removed this since now we're loading tenant/provider from request context and calling the provider's `RefreshSession` and `ValidateSession` methods.

JoelSpeed commented on tenant/config.go:
```diff
@@ -0,0 +1,12 @@
+package tenant
```
Why is this package not under `pkg` like the rest of the code?

salmanazmat666 replied:
will move to under `pkg/`

salmanazmat666 replied:
resolved, moved the new packages under the `pkg/` folder

JoelSpeed commented on tenant/configloader/config.go:
```diff
@@ -0,0 +1,22 @@
+	Rules       []*RuleConfig
+}
+
+// structure that is loaded from the tenant's config file
```
Oh wait a minute, so, once you have your provider configured, this adds a new dimension of tenants within the provider? This is very keycloak esque and doesn't translate well to other providers

salmanazmat666 replied:
Rules are what match an incoming request to a tenant/provider. The ConfigLoader configures all the providers and keeps them in memory. Then in a middleware (`pkg/middleware/tenantLoader.go`), the request is matched to a tenant and then that tenant is injeted into the request context. The rules are there to do the matching. We execute rules one by one untill we find a match.
A rule (`type RuleConfig struct`) has following structure.

```
type RuleSource string

const (
   SourceHost        RuleSource = "host"
   SourcePath        RuleSource = "path"
   SourceQueryParams RuleSource = "query"
   SourceHeader      RuleSource = "header"
)

type RuleConfig struct {
  Source       RuleSource // which part of the HTTP request contains the tenant id
  Expr         string     // the regex expression to match and extract tenant id from the source
  CaptureGroup int        // capture group or sub-match that is actually the tenant id
  QueryParam   string     // query parameter in case source is 'query'
  Header       string     // header key in case source is 'header'
}
```

I think these rules are pretty generic and are not just related to keycloak. We just have to have tenant-id somewhere in the incoming request. That location is specified the `Source` field in a rule. Right now, host, path, header, and query are supported but we can expand to include cookies as well.

JoelSpeed commented on tenant/configloader/loader.go:
```diff
@@ -0,0 +1,112 @@
+func New(conf *Configuration) (*loader, error) {
+	tntsConf := &TenantsConfig{}
+
+	err := loadConfigFromFile(conf.TenantsFile, tntsConf)
```
Sharding of config files is just going to make our configuration story harder IMO

salmanazmat666 replied:
will merge the config files,

salmanazmat666 replied:
resolve, only once config file for now.

JoelSpeed reviewed (commented):
I don't think we should mix `tenants` and `providers` up like this, right now, we have a single `provider` configuration, so the intention from previous work was that you would then have a list of `providers`, which is kind of what you've done, but you're calling it `tenants`. I'd rather we stick to a single term than mix these up.

I think this has the ground work needed, but how does it relate to existing PRs that are open? Is this an extension of those? Is it intended to replace them?

salmanazmat666 commented:
> I don't think we should mix `tenants` and `providers` up like this, right now, we have a single `provider` configuration, so the intention from previous work was that you would then have a list of `providers`, which is kind of what you've done, but you're calling it `tenants`. I'd rather we stick to a single term than mix these up.

thanks for the comment,
i can remove the "tenant" construct, so now following two new constructs will be introduced,

```
type TenantMatcher interface{
  GetTenantId(req *http.Request) (tenantId string, err error)
} 

type ProviderLoader interface{
// for now 'providerId' won't be used since each tenant will have only one provider but once PR #1657 is merged 'providerId' will be used
  Load(tenantId string, providerId string) (providers.Provider, error) 
}
```

so in a middleware, we can use TenantMatcher to get the tenantId out from http.Request, and then ProviderLoader to load the provider and inject it into request Context, so that it can be used everywhere where it is needed. What are your thoughts ?

> how does it relate to existing PRs that are open? Is this an extension of those? Is it intended to replace them?

I can't find any open PRs that implement multi-tenancy. #1657 will have conflicts with this PR, but I feel they won't be huge.

Commit:
incorporate PR comments
- remove tenant structure,
- just one config file for everything, no configuration sharding anymore,
- bring the moved files back to their original location,
- introduce tenantMatcher construct and tenantMatcher middleware that injects tenant-id in context
- replace tenantLoader with providerLoader

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

Commit:
inject tenantId in redirected requests

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

Commit:
Add tenantId in sessionState and create cookie names based on tenantid

Signed-off-by: Salman Azmat <salman.azmat@intechww.com>

salmanazmat666 commented:
Hi @JoelSpeed,

I have updated the PR to incorporate your comments. Following changes were made

1. Removed the `TenantLoader` interface. Now we have following constructs,
```
// tenantMatcher is responsible for matching tenants against incoming requests
type Matcher struct {}

// this function extracts tenantId from the request and returns it
func (matcher *Matcher) Match(req *http.Request) string {}

// ProviderLoader is responsible for getting Provider given a tenantId
type Loader interface {
	// id is provider id, which should be same as tenantId
	Load(id string) (providers.Provider, error)
}
```
2. Besides above constructs, I added two middlewares, one that extracts tenant-id from incoming request and injects it into request context, second that loads provider and injects it into request context.
3. Moved the new packages under the `pkg/` folder.
4. There is only one configuration file now, no more configuration sharding.
5. Added tenant-id as part of `SessionState`.
6. Implemented a decorator/wrapper over Sessionstore interface, that inserts tenant-id in the `Save(http.ResponseWrite, *http.Request, *SessionState) error` function and validates that the tenant-id in the `SessionState` and the incoming requests same inside the `Load(*http.Request) (*SessionsState, error)` function.
7. Some options that were moved to different packages have been restored to original locations,
8. Cookie names now have hash of tenant-id suffixed to them. So that we log in to the different tenants from the same browser window.

Please let me know your comments.

axel7083 commented:
Hello @salmanazmat666  and @JoelSpeed,

I am very interesting in this feature, therefore to participate in the effort I opend two PR on the @salmanazmat666 repository.
- https://github.com/salmanazmat666/oauth2-proxy/pull/1 (Merging and start fixing tests)
- https://github.com/salmanazmat666/oauth2-proxy/pull/2 (Adding the support for the SkipJwtBearerTokens option).

Hope this can still be integrated.

Commit:
add context parameter to provider loader

Signed-off-by: shifa imran <shifa.imran@intechww.com>

salmanazmat666 committed:
verify tenantid in csrf cookie and incoming request

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
implement postgres(with redis cache) tenant loader

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
fix unit tests

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 commented:
Hey @JoelSpeed,

We have added more features into the PR,

1. fixed unit tests
2. implemented a new tenantloader, that stores provider configs in postgres and uses redis as a cache to improve on latency, this also lets us add/remove providers while the proxy is running, via a REST api,
3. added context parameter to providerLoader's Load function,
4. added tenantId in CSRF cookie,


You comments are highly appreciated, please have a look at the changes and let me know what you think. Since it's a large PR we can have a call to discuss to design.

Regards,

adarshsingh2 commented:
Tremendous Job @salmanazmat666 . Do you have any sample config file or instructions to run these changes? For eg.
- sample tenants.yml file
- How to pass Postgres credentials and specific configurations
- sample alpha-config having tenantLoader and providerLoader configured
I would appreciate it if you could provide the above details.

salmanazmat666 commented:
> Tremendous Job @salmanazmat666 . Do you have any sample config file or instructions to run these changes? For eg.
> 
> * sample tenants.yml file
> * How to pass Postgres credentials and specific configurations
> * sample alpha-config having tenantLoader and providerLoader configured
>   I would appreciate it if you could provide the above details.

Hi @adarshsingh2,
In the current state we don't need `tenants.yaml`, just need following alpha-config, the tenant-matcher in this config gets the tenant from a query parameter `?tid=tenant1`, and the providerLoader loads the providers stored in a postgres database and uses redis as a cache, you can configure the following file according to your environment, we have implemented a REST client as well, that you can use to add/remove providers to postgres at runtime, use `pkg/providerloader/postgres/NewAPIClient` to get the client struct,

```
injectResponseHeaders:
- name: X-Forwarded-Access-Token
  values:
  - claim: access_token
- name: Authorization
  values:
  - claim: id_token
    prefix: 'Bearer '
- name: X-Auth-Request-User
  values:
  - claim: user
- name: X-Auth-Request-Email
  values:
  - claim: email
- name: X-Auth-Request-Preferred-Username
  values:
  - claim: preferred_username
- name: X-Auth-Request-Groups
  values:
  - claim: groups
- name: X-Auth-Request-Access-Token
  values:
  - claim: access_token
metricsServer:
  BindAddress: ""
  SecureBindAddress: ""
  TLS: null
providers:
  - id: tenant3
    provider: "keycloak"
    profileURL: "http://localhost:8080/auth/realms/tenant3/protocol/openid-connect/userinfo"
    validateURL: "http://localhost:8080/auth/realms/tenant3/protocol/openid-connect/userinfo"
    clientID: cmms
    clientSecret: "hf39jrh93uhd93wjd4iwj"
    scope: "openid email profile"
    loginURL: http://localhost:8080/auth/realms/tenant3/protocol/openid-connect/auth
    loginURLParameters:
    - default:
      - force
      name: approval_prompt
    oidcConfig:
      audienceClaims:
      - aud
      emailClaim: email
      groupsClaim: groups
      insecureAllowUnverifiedEmail: true
      insecureSkipIssuerVerification: true
      insecureSkipNonce: true
      issuerURL: http://localhost:8080/auth/realms/tenant3
      jwksURL: http://localhost:8080/auth/realms/tenant3/protocol/openid-connect/certs
      skipDiscovery: true
      userIDClaim: email
    redeemURL: http://localhost:8080/auth/realms/tenant3/protocol/openid-connect/token
server:
  BindAddress: 0.0.0.0:4180
  SecureBindAddress: ""
  TLS: null
tenantMatcher:
  rules:
    - source: "query"
      queryParam: "tid"
      expr: '.*'
      captureGroup: 0
providerLoader:
  type: "postgres"
  postgresLoader:
    postgres:
      host: "localhost"
      port: 5432
      database: "oauth2-proxy"
      schema: "oauth2_proxy"
      user: "admin"
      password: "admin"
      maxConnections: 50
      sslMode: "disable"
      sslRootCert: ""
      sslCrl: ""
      sslCert: ""
      sslKey: ""
    redis:
      connectionURL: "redis://localhost:6379"
      password: ""
      useSentinel: false
      sentinelPassword: ""
      sentinelMasterName: ""
      sentinelConnectionURLs:
      useCluster: false
      clusterConnectionURLs:
      cAPath: ""
      insecureSkipTLSVerify: false
      idleTimeout: 60
      expiry: 60000000000
      prefix: "oauth2-proxy-"
    aPI:
      host: "0.0.0.0"
      port: 8080
      pathPrefix: ""
      timeout: 60000000000
```

imranshifa committed:
Ci 304 resolve code climate failed checks (#5)

* feat: CI-304 resolve code climate issues 

Signed-off-by: “imranshifa” <“shifaimran750@gmail.com”>

---------

Signed-off-by: “imranshifa” <“shifaimran750@gmail.com”>
Co-authored-by: “imranshifa” <“shifaimran750@gmail.com”>

salmanazmat666 committed:
fix duplicate code issue

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

imranshifa committed:
fix: CI-313 add qoutes around schema name (#7)

Signed-off-by: “imranshifa” <“shifaimran750@gmail.com”>
Co-authored-by: “imranshifa” <“shifaimran750@gmail.com”>

imranshifa committed:
feat: CI-316 Add JWT parsing capability in tenant matcher middleware (#6)

Signed-off-by: “imranshifa” <“shifaimran750@gmail.com”>
Co-authored-by: “imranshifa” <“shifaimran750@gmail.com”>

imranshifa committed:
Ci 321 provider configs in postgres should be encrypted (#8)

Commit:
fix: remove empty ssl specifications from prostgres connection string

Signed-off-by: Shifa Imran <shifa.imran@sensys.com>

Commit:
fix: CI-369 decode jwt token properly to extract tenantid

Signed-off-by: Shifa Imran <shifa.imran@sensys.com>

OmarHawk commented:
> This pull request has been inactive for 60 days. If the pull request is still relevant please comment to re-activate the pull request. If no action is taken within 7 days, the pull request will be marked closed.

@JoelSpeed 
So sad about how much work has been spent on this. Maybe you should put somewhere a hint in the readme, that this repo won't be maintained anymore instead of people letting waste plenty of hours into Pull Request that will never be merged due to capacity-0% reasons.

ceastman-r7 commented:
this repo is no longer maintained?

tuunit commented:
Hi @salmanazmat666, Hi @OmarHawk,

first of all, your work is appreciated and will not be waste!

I recently joined Joel in the effort to get this project back on a smoother track and as you can see we released v7.5.0 last week and will be releasing v7.5.1 next Friday. Furthermore, a "help wanted" issue has been open for quite some time. https://github.com/oauth2-proxy/oauth2-proxy/issues/1848

I know that doesn't help to much with the fact that this PR has be dormant for a while. I will try my best to give it a new review soon.

tuunit commented:
Hi @salmanazmat666, to make up for my promise of giving this a proper review. I want to ask if you have this version of oauth2-proxy deployed in an environment of yours or not? Either way, would you be up for a meeting next week or so, to go through your implementation and architectural decisions together? Feel free to contact me via email or LinkedIn.

I'll start going through the code and try to find time to set it up on a test system.

salmanazmat666 commented:
> Hi @salmanazmat666, to make up for my promise of giving this a proper review. I want to ask if you have this version of oauth2-proxy deployed in an environment of yours or not? Either way, would you be up for a meeting next week or so, to go through your implementation and architectural decisions together? Feel free to contact me via email or LinkedIn.
> 
> I'll start going through the code and try to find time to set it up on a test system.

Hello @tuunit , 
Thanks for your message. I'll get in touch with you early next week for setting up the time of meeting. And about the deployment, yes, we are using this feature in production.

segaura commented:
Are there any update on this?
Did you managed to meet?

This is an essential feature for us (me and my company) so we are extremely interested and we are also willing to help.
We consider the currently proposed workaround of instantiating more than one pool of oauth2-proxies a too complex solution.

(**humble mode on**)
Also I think it'll be great if we find a way to:
- elect an issue, e.g. this one, and mark all other related issues and PRs as dead and link this one to them
- define the roadmap, e.g. if it is possible decide that v7 can be targeted, otherwise consider including this as a requirement in [v8 roadmap](https://github.com/oauth2-proxy/oauth2-proxy/projects/6)
- define a plan, e.g. deliverables/steps/milestones to reach the goal and corresponding dates

tuunit commented on pkg/tenant/matcher/matcher.go:
```diff
@@ -0,0 +1,121 @@
+package tenantmatcher
```
I think we should merge everything inside `pkg/tenant` into `pkg/providers`

tuunit replied:
Furthermore, I would suggest everything inside `pkg/providerloader` should be merged into a subpackage of providers: `pkg/providers/loader`

tuunit commented:
Hi @segaura and all other interested parties,

@salmanazmat666 and I had a meeting a few days ago, which was quite insightful.

Important Todo's for getting this PR ready:
- [ ] Refactoring of naming schema: tenants -> providers
- [ ] Refactoring of package structure `tenant` -> `providers`, `providerloader` -> `providers/loader`
- [ ] Fixing all existing unit tests
- [ ] Check context handling in `sessions.go > validateRedisSessionStore(o *options.Options)`
- [ ] Remove support for legacy (toml) flags and config. As the old config will be deprecated soon, we should only concern ourselves with the new yaml based config in this PR.

I'm suggestion the following:
This feature branch should be kept to a bare minimum of supported features. Therefore support for multiple providers will not be implemented as part of this PR for the following features:
  - Email domains and authenticated emails validator
  - Trusted IPs validator
  - Basic Auth validator
  - SkipJwtBearerTokens


Before going ahead with this PR, I would like to introduce a new structured logging implementation including log levels and #2279 for full support of the structured configuration (yaml, currently called alpha config) first (v7.6 and v8). I will try my best to get this PR in with v8.1 (Most likely somewhere around march - may)

salmanazmat666 committed:
chore: update redislock version

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
Merge remote-tracking branch 'upstream/master' into feature/multi-tenancy

salmanazmat666 committed:
refactor: rename tenant to provider

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
refactor: move package pkg/providerLoader to providers/loader

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
feat: templatize cookie domains

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
fix: fix state encode/decdode logic

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
Merge branch 'master' into feature/multi-tenancy

salmanazmat666 committed:
fix: fix number of states

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
tests: fix unittests and linting

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
refactor: refactor backendLogout() to have < 4 returns

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

tuunit commented:
@salmanazmat666 happy to see you back at it

salmanazmat666 committed:
refactor: refactor encryption decorator

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
changed comment and return type

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
Merge remote-tracking branch 'origin/master' into feature/multi-tenancy

salmanazmat666 committed:
feat: add local environment files

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 committed:
Merge remote-tracking branch 'upstream/master' into feature/multi-tenancy

salmanazmat666 commented:
@tuunit I have updated the PR and following are the main things I did,

- [ ] - Refactoring of naming schema: tenants -> providers
- [ ] - Refactoring of package structure tenant -> providers, providerloader -> providers/loader
- [ ] - Fixing all existing unit tests
- [ ] - Check context handling in sessions.go > validateRedisSessionStore(o *options.Options)
- [ ] - Remove support for legacy (toml) flags and config. We still have necessary config in legacy config file like cookie settings etc, so cannot remove that unless we migrate all configs to alpha.

Besides, I have also created files for local environment setup to demo the multi-tenancy feature, we can now use `make multi-tenancy-up` to bring up the necessary services that are already configured for multi-tenancy support.

Please have a look at the changes and let me know if you want any other change.

Regards,

tuunit commented on contrib/local-environment/Makefile:
```diff
@@ -14,6 +14,14 @@ alpha-config-up:
 alpha-config-%:
 	docker compose -f docker-compose.yaml -f docker-compose-alpha-config.yaml $*
 
+.PHONY: multi-tenancy-up
```
can we rename this to:

```suggestion
.PHONY: multi-provider-up
```

tuunit commented on contrib/local-environment/docker-compose-multi-tenancy.yaml:
and rename these files and replace multi-tenancy with multi-provider

tuunit reviewed (changes requested):
I'll try to do a full review again soon and do some local testing as well.

tuunit commented:
Thanks for putting all the work in 🥳

salmanazmat666 committed:
chore: rename tenant -> provider

Signed-off-by: Salman Azmat <salman.azmat@sensys.com>

salmanazmat666 commented:
> Thanks for putting all the work in 🥳

Hey @tuunit , did you get a chance to review the PR ?

robertlagrant commented:
@salmanazmat666 it might be worth making progress by merging in the base branch?

alexdutton commented:
Hi @salmanazmat666. I'm interested in this feature too, and thank you for all your work so far!

I've given it a go, and wanted to check my understanding.

1. At the moment, am I right in thinking that every request (not just within the authentication flow) needs to be mapped to a provider, otherwise oauth2-proxy will return a 401?
2. There's currently a `config` provider loader and a `single` provider loader — could this be simplified to just the `config` loader with an implicit default of the first provider?

Our use case is to support multiple providers for a single upstream service, all from a single public service URL. It'd be good (at least from our perspective) if the provider can be specified as a query parameter to `/oauth2/start`/`oauth2/sign_in` and then for the provider ID to be held in OIDC state and then the session. We only protect our API's login endpoint, so requests to all other paths are unauthenticated.

I've started experimenting with adapting what you've got so far to implement the suggestion above, but don't want to go off in another direction to where you're headed. If you'd like to catch up on this that'd be great!

All the best, ~ Alex

salmanazmat666 committed:
fix: url encode-decode redirect in state

Signed-off-by: salmanazmat666 <salman.azmat@sensys.com>

salmanazmat666 committed:
fix: fix needsRefresh to check for token expiry as well

Signed-off-by: salmanazmat666 <salman.azmat@sensys.com>

stleon commented:
Thank you for this amazing product and all the incredible work you've done — it's truly impressive! 🙌

Regarding multi-tenancy functionality: quite some time has passed since this idea was discussed, and it would be fantastic to see it implemented. Are there any updates on this topic?

salmanazmat666 committed:
Merge branch 'master' into feature/multi-tenancy

Um-e-Hani committed:
chore: update unittests

Signed-off-by: Um E Hani <ume.hani@sensys.com>

salmanazmat666 committed:
chore: update go-redis to v9.7.1

Signed-off-by: salmanazmat666 <salman.azmat@sensys.com>

salmanazmat666 commented:
@tuunit I see you have added `high-priority` label on this PR.

We have also recently merged the main branch in this branch.

Let me know when you want to move forward with merging this PR.

rashidriaz committed:
chore: CI-1103 update viper version from v1.19.0 to v1.21.0 (#28)

rashidriaz committed:
feat: CI-1455 sync oAuth2 proxy with upstream repo (#29)

* deps: remove tools/reference-gen from go.mod

Remove github.com/oauth2-proxy/tools/reference-gen from dependencies.
Instead we are now running it with "go run" with a version suffix.

Long version:
- github.com/oauth2-proxy/tools/reference-gen is removed from
  tools/tool/go
- in pkg/apis/options/doc.go we now run reference-run with a version
  suffix (go run package@version) with the version comming from go.mod.
- the "//go:generate" line is split in 2 lines (using the -command
  flag) for readability
- "go mod tidy" for cleaning dependencies from go.mod, go.sum

Note: we are not upgrading reference-gen here. That will be a further
separate change.

* chore(deps): update dependency @easyops-cn/docusaurus-search-local to ^0.49.0

* chore(deps): update gomod (#2952)

* chore(deps): update gomod

Co-authored-by: Jan Larwig <jan@larwig.com>

* chore(deps): update module github.com/golang-jwt/jwt/v5 to v5.2.2 [security] (#3003)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update golang dependencies and pin to latest golang v1.23.x release (#3011)

Signed-off-by: Jan Larwig <jan@larwig.com>

* release v7.8.2 (#3012)

* update to release version v7.8.2

* docs: release letter

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* chore(deps): update dependency golangci/golangci-lint to v1.64.8 (#3004)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* feat: update HashNonce to use crypto/sha256 (#2967)

Signed-off-by: egibs <20933572+egibs@users.noreply.github.com>

* chore(deps): update docker-compose (#3005)

* docs: fix gitlab docs url for oauth2 integration (#3002)

* Update gitlab.md with correct url for creating an application

* docs: fix gitlab docs url for oauth2 integration

---------

Co-authored-by: Jan Larwig <jan@larwig.com>

* feat: allow to set non-default authorization request response mode (#3001)

* Update Go version in devcontainer

* Add option to change response mode in authorization request

* Fix option name

* Update docs and changelog

* Rename config value to underscore

* Add unit tests for added parameter

* Move change to upcoming release

* Generate alpha config

---------

Co-authored-by: Michael Cornel <michael@stieler.it>

* chore(deps): update docker-compose

* chore(deps): update gitea/gitea docker tag to v1.23.7 (#3030)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* Add --bearer-token-login-fallback option (#2924)

* add --deny-invalid-bearer-tokens

* update changelog

* PR feedback, update api-routes description

* update --api-routes description

* revert load_test fix that I needed locally

---------

Co-authored-by: Justin Ryan <j.ryan@mwam.com>

* chore(deps): update module golang.org/x/net to v0.38.0 [security] (#3035)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): upgrade to latest golang v1.23.x release (#3041)

* chore(deps): update docker-compose (#3038)

* fix(entra-id): use federated credentials for refresh token (#3031)

* fix: use federated credentials to refresh token in entra id

* fix: add some error handling

* chore: update changelog

* chore: update comments

* chore: update comments

* doc: reference entra id docs and clearer phrasing of comments

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* chore(deps): update ghcr.io/dexidp/dex docker tag to v2.42.1 (#3044)

* doc(entra-id): correct toml field in sample (#2946)


---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* chore(build): refactoring makefile for better usability and introducing a default help target (#2930)

* fix: role extraction from access token in keycloak oidc  (#1916)

* Fix wrong token used in Keycloak OIDC provider

* Update CHANGELOG for PR #1916

* Update tests

* fix: keycloak oidc role extraction

---------

Co-authored-by: Jan Larwig <jan@larwig.com>

* feat: ability to parse JWT encoded profile claims (#3014)

* fix: parse JWT profile claims

* Comment with OIDC specs reference

* fix: formatting

* Updated changelog

---------

Co-authored-by: Jan Larwig <jan@larwig.com>

* release: v7.9.0 (#3047)

* add new docs version 7.9.x

* update to release version v7.9.0

* doc: add changelog summary

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* docs: clear up multiple-providers is unimplemented (#3046)

* docs: clear up multiple-providers is unimplemented

Currently this configuration option is held up by #926. So users don't
assume this solution will work for them, and later find the feature is
not yet implemented -- own the shortcoming clearly.

* doc: add note about missing multi provider implementation to versioned docs

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* docs: add note about version obfuscation to footer option (#3051)

Signed-off-by: Jan Larwig <jan@larwig.com>

* chore(deps): update docker-compose (#3074)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): drop golang.org/x/exp/{slices,maps} (#3065)

Use Go 1.23 stdlib instead: slices, maps.

* chore(lint): fix staticcheck issues (#3061)

* chores: fix staticcheck QF1012

Fix use of fmt.Sprintf when writing to a writer.
https://staticcheck.dev/docs/checks/#QF1012
https://github.com/oauth2-proxy/oauth2-proxy/issues/3060

* chores: fix staticcheck QF1003

Use switch instead of multiple if/else.
https://staticcheck.dev/docs/checks/#QF1003
https://github.com/oauth2-proxy/oauth2-proxy/issues/3060

* chores: exclude staticcheck QF1008 for now

We aim to migrate golangci-lint to v2
Let's disable QF1008 (Omit embedded fields from selector expression)
for now.
https://staticcheck.dev/docs/checks/#QF1008

* chores: fix golangci config: run.deadline -> timeout

Rename config option to match v1 documentation: deadline -> timeout.
https://golangci.github.io/legacy-v1-doc/usage/configuration/#run-configuration

This error has been spotted by golangci-lint v2 migration tool.

* chores: fix staticcheck QF1012

* chores: major upgrade of golangci-lint from v1.64.8 to v2.1.6 (#3062)

.golangci.yml migrated with "golangci-lint migrated" and then tweaked to add comments back.

* feat: support for multiple github orgs (#3072)

* fix for github teams

* Update github.go

* added errorhandling

* Update github.md

* refactored GitHub provider

refactored hasOrg, hasOrgAndTeams and hasTeam into hasAccess to stay within function limit

* reverted Refactoring

* refactored github.go

- joined hasOrgAndTeamAccess into checkRestrictions

* refactored github.go

- reduced number of returns of function checkRestrictions to 4

* updated GitHub provider to accept legacy team ids

* GoFmt and golangci-lint

Formatted with GoFmt and followed recommendations of GoLint

* added Tests

added Tests for checkRestrictions.

* refactored in maintainer feedback

* Removed code, documentation and tests for legacy ids

* add changelog and update docs

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* Create FUNDING.yml

* feat: bump to go1.24.5 and full dependency update (#3116)

* upgrade to go1.24.5

dependency updates

lint fixes

chore(deps): upgrade github.com/spf13/viper to v1.20.1

Note that this upgrade also implied to upgrade github.com/mitchellh/mapstructure
(nowadays unmaintained: https://gist.github.com/mitchellh/90029601268e59a29e64e55bab1c5bdc)
to github.com/go-viper/mapstructure/v2.

fix: adapt tests to match mapstructure v2 error messages

pkg/apis/options/load_test.go: skip tests on Go 1.23

Add a compile guard for Go < 1.24 for the pkg/apis/options/load_test.go
because the LoadYAML test depends on error messages produced by
encoding/json that changed slightly (names of embedded structs are now
reported). As we updated the test for go1.24, the test now fails on
1.23, but just for a slight difference, so we disable the test there.

fix: adapt tests to match mapstructure v2 error messages

remove pre 1.24 disclaimer

add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

Co-Authored-By: Olivier Mengué <dolmen@cpan.org>

* add exclusion for 'avoid meaningless package names' in .golangci.yml

* chore(dep): upgrade all dependencies

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Olivier Mengué <dolmen@cpan.org>
Co-authored-by: Jan Larwig <jan@larwig.com>

* chore(deps): update dependency golangci/golangci-lint to v2.2.2 (#3111)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update dependency @easyops-cn/docusaurus-search-local to ^0.51.0 (#3098)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update example docker-compose files (#3096)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update helm examples (#2951)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update alpine base image to v3.22.0 (#3097)

* chore(deps): update alpine docker tag to v3.22.0

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* fix: return error for empty Redis URL list (#3101)

* fix: return error for empty Redis URL list

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* add new docs version 7.10.x

* update to release version v7.10.0

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

* chore(deps): update dependency @easyops-cn/docusaurus-search-local to ^0.52.0 (#3131)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update gomod (#3132)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update helm release oauth2-proxy to v7.14.1 (#3133)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update docker-compose (#3130)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update alpine docker tag to v3.22.1 (#3129)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* feat(cookie) csrf per request limit (#3134)

* Allow setting maximum number of csrf cookies, deleting the oldest if necessary

* Add a test for multiple CSRF cookies to remove the old cookie

* Add docs/changelog

* If limit is <=0 do not clear

Signed-off-by: test <bert@transtrend.com>

* Better docs

Co-authored-by: Jan Larwig <jan@larwig.com>

* direct check of option value

Co-authored-by: Jan Larwig <jan@larwig.com>

* direct use of option value

Co-authored-by: Jan Larwig <jan@larwig.com>

* sort based on clock compare vs time compare

Co-authored-by: Jan Larwig <jan@larwig.com>

* clock.Clock does not implement Compare, fix csrf cookie extraction after rename

Signed-off-by: Bert Helderman <bert@transtrend.com>

* Linter fix

* add method signature documentation and slight formatting

Signed-off-by: Jan Larwig <jan@larwig.com>

* fix: test case for csrf cookie limit and flag

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Bert Helderman <bert@transtrend.com>
Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: test <bert@transtrend.com>
Co-authored-by: bh-tt <71650427+bh-tt@users.noreply.github.com>

* fix: show login page on broken session cookie (#2605)

* fix: redirect on invalid cookie

* docs: update changelog

* chore: remove duplicated code

* fix: status code handling if wrong http method is used

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* Fix local-environment ports (#3136)

* Change Dex port in local-environment from 4190 to 5556

Port 4190 is blocked by standards-compliant browsers (e.g. Firefox), as per https://fetch.spec.whatwg.org/#port-blocking.
Port 5556 is used by Dex in its example config files: https://github.com/dexidp/dex/blob/745e1114f341e849f3b0edde45b39c14017deaf8/examples/config-dev.yaml#L50

* Fix upstream in local-environment/oauth2-proxy.cfg

http://httpbin.localtest.me:8080 is only exposed to the host, not to httpbin Docker network.
Causes Bad Gateway before.

* Do not expose unauthenticated httpbin service in local-environment

This defeats the point of having oauth2-proxy.
It has already been misleading by causing the bug fixed in cafc6af48fc38f6fe4395fb0c7e2638bc84e6091.
It serves as a bad example: users might accidentally expose the service they're trying to protect in the first place.

* Remove unnecessary httpbin.localtest.me alias from local-environment

* feat: allow use more possible google admin-sdk api scopes (#2743)

* feat: Allow use more possible google admin-sdk api scopes.

* reduce cognitive complexity

Signed-off-by: Bob Du <i@bobdu.cc>

* remove unnecessary else block / indentation

Signed-off-by: Jan Larwig <jan@larwig.com>

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

* slight formatting and error message rephrasing

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Bob Du <i@bobdu.cc>
Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* feat: add SourceHut (sr.ht) provider (#2359)

* Add SourceHut (sr.ht) provider

* fix changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* fix: regex substitution for $ signs in upstream path handling before running envsubst (#2524)

* Perform a regex replace of $NUM to $$NUM before running envsubst

* Perform a regex replace of $NUM to $$NUM before running envsubst

* add test case; fix linter warnings; add method documentation

Signed-off-by: Jan Larwig <jan@larwig.com>

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* feat(cookie): add feature support for cookie-secret-file (#3104)

* feat: add feature support for cookie-secret-file

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-Authored-By: Sandy Chen <Yuxuan.Chen@morganstanley.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* feat: use non-default authorization request response mode in OIDC providers (#3055)

* fix: OIDC sets response mode

* Update CHANGELOG

* feat: make google-groups argument optional (#3138)

add test cases

update documentation

refactor code and some cleanup

update changelog

Signed-off-by: Jan Larwig <jan@larwig.com>

* feat: differentiate between "no available key" and error for redis sessions (#3093)

* add some better error handling

* add changelog entry

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* Merge commit from fork

Signed-off-by: Jan Larwig <jan@larwig.com>

* release v7.11.0 (#3145)

* add new docs version 7.11.x

* update to release version v7.11.0

* add changelog entry for v7.11.0

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* fix: port for local-environment (#3148)

* feat(e2e): add workflow to trigger e2e test suite through PR comments (#3153)

* feat(e2e): add workflow to trigger e2e test suite through PR comments

* add empty line

* feat: add Cidaas provider (#2273)

* Add sensible logging flag to default setup for logger

* Fix default value flag for sensitive logging

* Remove sensitive logging changes

* Add Cidaas provider

* Update CHANGELOG.md

* Add required groups scope to defaults

* Fix tests

* Remove if block with protected resource

* Fix linting

* Adjust provider sorting, fixes

* Directly handle error return

Co-authored-by: Jan Larwig <jan@larwig.com>

* Use less deep nesting

Co-authored-by: Jan Larwig <jan@larwig.com>

* Directly handle returned error

Co-authored-by: Jan Larwig <jan@larwig.com>

* Pass provider options to Cidaas provider

Co-authored-by: Jan Larwig <jan@larwig.com>

* Add import for provider options

* Fix tests

* Fix linting

* Add Cidaas doc page

* Add Cidaas provider doc page to overview

* Fix link in docs

* Fix link in docs

* Add link to Cidaas

* fix provider order in docs and changelog position

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Teko012 <112829523+Teko012@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Kevin Kreitner <kevinkreitner@gmail.com>

* chore(dep): upgrade to latest golang 1.24.6 (#3166)

Signed-off-by: Jan Larwig <jan@larwig.com>

* chore(deps): update actions/checkout action to v5 (#3164)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update dependency golangci/golangci-lint to v2.4.0 (#3161)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update docker-compose (#3160)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* feat: allow disable-keep-alives configuration in upstream (#3156)

Signed-off-by: Jan Larwig <jan@larwig.com>

* fix: Gitea team membership (#3150)

* bugfix: Gitaa team membership

Gitea doesn't properly fill in all the fields like GitHub,
so implement a series of fallbacks.

Signed-off-by: magic_rb <magic_rb@redalder.org>

* add changelog, documentation and fix groups list

Signed-off-by: Jan Larwig <jan@larwig.com>

---------

Signed-off-by: magic_rb <magic_rb@redalder.org>
Signed-off-by: Jan Larwig <jan@larwig.com>
Co-authored-by: Jan Larwig <jan@larwig.com>

* add new docs version 7.12.x

Signed-off-by: Jan Larwig <jan@larwig.com>

* update to release version v7.12.0

* add changelog entry for v7.12.0

Signed-off-by: Jan Larwig <jan@larwig.com>

* doc: SourceHut documentation fixes (#3170)

* fix: SourceHut documentation

- Add it to sidebar and provider index
- Fix broken link

This fixes an oversight in #2359, where I had not fully understood how
the documentation works.

Signed-off-by: Conrad Hoffmann <ch@bitfehler.net>

* fix: doc build instructions in docs/README.md

---------

Signed-off-by: Conrad Hoffmann <ch@bitfehler.net>

* chore(deps): update actions/upload-pages-artifact action to v4 (#3194)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* doc: update contribution guide to avoid a specific mention of the version of Go being used. (#3157)

* fix(deps): revert actions/upload-pages-artifact action to v3 (#3211)

Signed-off-by: Richard Ahlquist <richard.jimmy.johansson@gmail.com>

* chore(deps): update alpine docker tag to v3.22.2 (#3241)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update actions/setup-node action to v6 (#3242)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update actions/stale action to v10 (#3193)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update actions/setup-go action to v6 (#3191)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update actions/labeler action to v6 (#3190)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update helmv3 (#3189)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update docker-compose (#3188)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update dependency golangci/golangci-lint to v2.5.0 (#3212)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore: CI-1455 remove comments

* chore: CI-1455 update docs/configs/alpha-configs.yaml after generate

* fix: CI-1455 lint

---------

Signed-off-by: Jan Larwig <jan@larwig.com>
Signed-off-by: egibs <20933572+egibs@users.noreply.github.com>
Signed-off-by: Bert Helderman <bert@transtrend.com>
Signed-off-by: Bob Du <i@bobdu.cc>
Signed-off-by: magic_rb <magic_rb@redalder.org>
Signed-off-by: Conrad Hoffmann <ch@bitfehler.net>
Signed-off-by: Richard Ahlquist <richard.jimmy.johansson@gmail.com>
Co-authored-by: Olivier Mengué <dolmen@cpan.org>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jan Larwig <jan@larwig.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Evan Gibler <20933572+egibs@users.noreply.github.com>
Co-authored-by: Copolycube <kaourintin+github@gmail.com>
Co-authored-by: Michael Cornel <michael@stieler.it>
Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>
Co-authored-by: Justin Ryan <jryan@verticalresponse.com>
Co-authored-by: Justin Ryan <j.ryan@mwam.com>
Co-authored-by: Enrico <enrico.pelizzon@gmail.com>
Co-authored-by: Richard Hagen <richard.hagen@gmail.com>
Co-authored-by: Michael Niksa <miniksa@microsoft.com>
Co-authored-by: Guillaume "Elektordi" Genty <github@elektordi.net>
Co-authored-by: ikarius <fred@ikarius.com>
Co-authored-by: Evan Carroll <me@evancarroll.com>
Co-authored-by: Daniel Mersch <94058351+daniel-mersch@users.noreply.github.com>
Co-authored-by: Edward Viaene <ward.viaene@gmail.com>
Co-authored-by: Daniel Givens <1581675+dgivens@users.noreply.github.com>
Co-authored-by: test <bert@transtrend.com>
Co-authored-by: bh-tt <71650427+bh-tt@users.noreply.github.com>
Co-authored-by: Johann <76482511+Primexz@users.noreply.github.com>
Co-authored-by: Simmo Saan <simmo.saan@gmail.com>
Co-authored-by: Bob Du <i@bobdu.cc>
Co-authored-by: Conrad Hoffmann <1226676+bitfehler@users.noreply.github.com>
Co-authored-by: Ashkan Daie <1415513+dashkan@users.noreply.github.com>
Co-authored-by: Sandy Chen <sandy19890604@gmail.com>
Co-authored-by: Sandy Chen <Yuxuan.Chen@morganstanley.com>
Co-authored-by: Sourav Agrawal <146818014+sourava01@users.noreply.github.com>
Co-authored-by: nobletrout <nobletrout@gmail.com>
Co-authored-by: Theron Boerner <hunterboerner@gmail.com>
Co-authored-by: Kevin Kreitner <kevin.kreitner@real-digital.de>
Co-authored-by: Teko012 <112829523+Teko012@users.noreply.github.com>
Co-authored-by: Kevin Kreitner <kevinkreitner@gmail.com>
Co-authored-by: jet <71936688+jet-go@users.noreply.github.com>
Co-authored-by: Richard Brežák <richard@brezak.sk>
Co-authored-by: David Symonds <dsymonds@gmail.com>
Co-authored-by: Richard Ahlquist <richard.jimmy.johansson@gmail.com>

mane-tv2 commented:
Any update on this ?
